daffodil-lib/src/main/scala/org/apache/daffodil/util/DataValue.scala

```diff
@@ -0,0 +1,91 @@
+  type DataValueCalendar = DataValue[DFDLCalendar, NonNullable with DataValuePrimitiveType]
+  type DataValueByteArray = DataValue[Array[Byte], NonNullable with DataValuePrimitiveType]
+  type DataValueBool = DataValue[JBoolean, NonNullable with DataValuePrimitiveType]
+  type DataValueNumber = DataValue[JNumber, NonNullable with DataValuePrimitiveType]  
+  type DataValueLong = DataValue[JLong, NonNullable with DataValuePrimitiveType]  
+  type DataValueDouble = DataValue[JDouble, NonNullable with DataValuePrimitiveType]  
+  type DataValueBigInt = DataValue[JBigInt, NonNullable with DataValuePrimitiveType] 
+  type DataValueString = DataValue[JString, NonNullable with DataValuePrimitiveType] 
+  type DataValueFloat = DataValue[JFloat, NonNullable with DataValuePrimitiveType]
+  type DataValueByte = DataValue[JByte, NonNullable with DataValuePrimitiveType]
+  type DataValueInt = DataValue[JInt, NonNullable with DataValuePrimitiveType]
+  type DataValueShort = DataValue[JShort, NonNullable with DataValuePrimitiveType]
+  type DataValueDINode = DataValue[DataValueDINodeTrait, NonNullable with DataValuePrimitiveType]
+
+  import scala.language.implicitConversions
+    
+  implicit def toDataValue(v: JBigDecimal): DataValueBigDecimal = new DataValue(v)
+  implicit def toDataValue(v: DFDLCalendar) : DataValueCalendar = new DataValue(v)
+  implicit def toDataValue(v: Array[Byte]) : DataValueByteArray = new DataValue(v)
+  implicit def toDataValue(v: JBoolean) : DataValueBool = new DataValue(v)
+  implicit def toDataValue(v: JNumber) : DataValueNumber = new DataValue(v)
+  implicit def toDataValue(v: JLong) : DataValueLong = new DataValue(v)
+  implicit def toDataValue(v: JDouble) : DataValueDouble = new DataValue(v)
+  implicit def toDataValue(v: JBigInt) : DataValueBigInt = new DataValue(v)
+  implicit def toDataValue(v: JString) : DataValueString = new DataValue(v)
+  implicit def toDataValue(v: JFloat) : DataValueFloat = new DataValue(v)
+  implicit def toDataValue(v: JByte) : DataValueByte = new DataValue(v)
+  implicit def toDataValue(v: JInt) : DataValueInt = new DataValue(v)
+  implicit def toDataValue(v: JShort) : DataValueShort = new DataValue(v)
+  implicit def toDataValue(v: DataValueDINodeTrait) : DataValueDINode = new DataValue(v)
+  
+  
+  implicit def toDataValue(v: Long) : DataValueLong = new DataValue(v:JLong)
+  implicit def toDataValue(v: Double) : DataValueDouble = new DataValue(v:JDouble)
+  implicit def toDataValue(v: Boolean) : DataValueBool = new DataValue(v:JBoolean)
+  implicit def toDataValue(v: Float) : DataValueFloat = new DataValue(v:JFloat)
+  implicit def toDataValue(v: Byte) : DataValueByte = new DataValue(v:JByte)
+  implicit def toDataValue(v: Int) : DataValueInt = new DataValue(v:JInt)
+  implicit def toDataValue(v: Short) : DataValueShort = new DataValue(v:JShort)
+
+  val NoValue = new DataValue(new NoValueObj)
+
+}
+
+protected class NoValueObj {
```

Comment:
If this is just a singleton, why not an object. Any why not make it  a private member of the DataValue object?

Reply:
Turned into an object. I cannot make it private, because "val NoValue = new DataValue(NoValueObj)" causes it to escape its defining scope.

Reply:
Disregard above. Making it an object prevents using it as a type, which means I cannot define the DataValueEmpty type.

I'm not convinced that we will ever actually use said type, but I don't see the harm of declaring it as a class that gets initialized to a singleton.

Reply:
No big deal. Just thought we could hide it better, but if it's problematic, don't bother.